source/slang/slang-check-decl.cpp

```diff
@@ -9308,8 +9308,161 @@ namespace Slang
             this, funcDecl, attr, DeclAssociationKind::PrimalSubstituteFunc);
     }
 
+    static int32_t findLevelsOfInheritance(AggTypeDecl* decl, int32_t inheritanceLevel, List<AggTypeDecl*>& linearInheritanceHierarchy)
+    {
+        // We need to assume multiple `InheritanceDecl` are possible per `decl` member list.
+        int32_t inheritanceDepth = inheritanceLevel;
+        for (auto maybeInheritance : decl->members)
+        {
+            auto inheritance = as<InheritanceDecl>(maybeInheritance);
+            if (!inheritance)
+                continue;
+            auto isValidInheritance = as<DeclRefType>(inheritance->base.type);
+            if (!isValidInheritance || !isValidInheritance->getDeclRef())
+                continue;
+            auto isInherited = as<AggTypeDecl>(isValidInheritance->getDeclRef().getDecl());
+            if (!isInherited)
+                continue;
+            linearInheritanceHierarchy.add(isInherited);
+            int32_t depthFound = findLevelsOfInheritance(isInherited, inheritanceLevel-1, linearInheritanceHierarchy);
+            if (depthFound < inheritanceDepth)
+                inheritanceDepth = depthFound;
+        }
+        return inheritanceDepth;
+    }
+
     void SemanticsDeclAttributesVisitor::visitStructDecl(StructDecl* structDecl)
     {
+        List<AggTypeDecl*> linearInheritanceHierarchy;
+        linearInheritanceHierarchy.add(structDecl);
+        int32_t levelsOfInheritance = findLevelsOfInheritance(structDecl, 0, linearInheritanceHierarchy);
+
+        List<ConstructorDecl*> ctorList;
+        bool fillCtorList = false;
+
+        auto findAllCtor = [&]() {
+            ConstructorDecl* ctorDefault = nullptr;
+            fillCtorList = true;
+            for (auto maybeCtor : structDecl->members)
+            {
+                auto ctor = as<ConstructorDecl>(maybeCtor);
+                if (!ctor)
+                    continue;
+                ctorList.add(ctor);
+
+                // It is possible BlockStmt has a child with the type of
+                // `ExpressionStmt` if an existing constructor has only 1
+                // expression. This would be a senario we need to
+                // put the `ExpressionStmt` inside a `SeqStmt`.
+                // Since we found an init expression, we need to ensure
+                // we are adding to a constructor which can handle multiple
+                // statments since we will be adding `init` expressions to
+                // the constructor.
+                auto block = as<BlockStmt>(ctor->body);
+                if (!as<SeqStmt>(block->body))
```

Comment:
similarly, this can be simplified as

if (tmpExpr = as<SeqStmt>(block->body))
{
                    auto seqStmt = m_astBuilder->create<SeqStmt>();
                    seqStmt->stmts.add(tmpExpr);
                    block->body = seqStmt;
}

Reply:
I can move the block scope definition into the if statement, I cannot remove the `!as<SeqStmt>(block->body)` check. The check ensures the as<BlockStmt>(ctor->body)->body is a `SeqStmt` to allow us to add more than 1 statement to the existing constructor for struct field init expression support.